chronos/asyncproc.nim

```diff
@@ -0,0 +1,1354 @@
+    stdinHolder: localStdin,
+    stdoutHolder: localStdout,
+    stderrHolder: localStderr,
+    stdinHandle: remoteStdin,
+    stdoutHandle: remoteStdout,
+    stderrHandle: remoteStderr
+  ))
+
+proc closeWait(holder: AsyncStreamHolder) {.async.} =
+  let (future, transp) =
+    case holder.kind
+    of StreamKind.None:
+      (nil, nil)
+    of StreamKind.Reader:
+      if StreamHolderFlag.Stream in holder.flags:
+        (holder.reader.closeWait(), holder.reader.tsource)
+      else:
+        (nil, holder.reader.tsource)
+    of StreamKind.Writer:
+      if StreamHolderFlag.Stream in holder.flags:
+        (holder.writer.closeWait(), holder.writer.tsource)
+      else:
+        (nil, holder.writer.tsource)
+
+  let pending =
+    block:
+      var res: seq[Future[void]]
+      if not(isNil(future)):
+        res.add(future)
+      if not(isNil(transp)):
+        if StreamHolderFlag.Transport in holder.flags:
+          res.add(transp.closeWait())
+      res
+
+  if len(pending) > 0:
+    await allFutures(pending)
+
+proc closeProcessStreams(pipes: AsyncProcessPipes,
+                         options: set[AsyncProcessOption]): Future[void] =
+  let pending =
+    block:
+      var res: seq[Future[void]]
+      if ProcessFlag.AutoStdin in pipes.flags:
+        res.add(pipes.stdinHolder.closeWait())
+      if ProcessFlag.AutoStdout in pipes.flags:
+        res.add(pipes.stdoutHolder.closeWait())
+      if ProcessFlag.AutoStderr in pipes.flags:
+        res.add(pipes.stderrHolder.closeWait())
+      res
+  allFutures(pending)
+
+proc closeWait*(p: AsyncProcessRef) {.async.} =
+  # Here we ignore all possible errrors, because we do not want to raise
+  # exceptions.
+  discard closeProcessHandles(p.pipes, p.options, OSErrorCode(0))
+  await p.pipes.closeProcessStreams(p.options)
+  discard p.closeThreadAndProcessHandle()
+  untrackAsyncProcess(p)
+
+proc stdinStream*(p: AsyncProcessRef): AsyncStreamWriter =
+  doAssert(p.pipes.stdinHolder.kind == StreamKind.Writer,
+           "StdinStreamWriter is not available")
+  p.pipes.stdinHolder.writer
+
+proc stdoutStream*(p: AsyncProcessRef): AsyncStreamReader =
+  doAssert(p.pipes.stdoutHolder.kind == StreamKind.Reader,
+           "StdoutStreamReader is not available")
+  p.pipes.stdoutHolder.reader
+
+proc stderrStream*(p: AsyncProcessRef): AsyncStreamReader =
+  doAssert(p.pipes.stderrHolder.kind == StreamKind.Reader,
+           "StderrStreamReader is not available")
+  p.pipes.stderrHolder.reader
+
+proc execCommand*(command: string,
+                  options = {AsyncProcessOption.EvalCommand},
+                  timeout = InfiniteDuration
+                 ): Future[int] {.async.} =
+  let poptions = options + {AsyncProcessOption.EvalCommand}
+  let process = await startProcess(command, options = poptions)
+  let res =
+    try:
+      await process.waitForExit(timeout)
+    finally:
+      await process.closeWait()
+  return res
+
+proc execCommandEx*(command: string,
```

Comment:
can you add the ability to pass arguments to startProcess?

Reply:
I'm not sure i understand your question what arguments you want to pass?

Reply:
Like this: https://github.com/status-im/nim-chronos/commit/7bbc59f448b4da737a0905efe0b855d22deccf01

Reply:
But `execCommandEx` will not support all the arguments for `startProcess`, so we need to create different subset of options for it.

Reply:
Oh i see you want `arguments` not `options`, but execCommand is supposed to start shell commands, so you can combine your options to a command with `command & arguments.join(" ")`

Reply:
> arguments.join

this would introduce quoting / whitespace issues, no?

Reply:
> But `execCommandEx` will not support all the arguments for `startProcess`, so we need to create different subset of options for it.

Is there any benefit from that? IMHO it looks like inconsistency in the API.

Reply:
It doesnt look like inconsistency of the API, because `execCommand` is supposed to invoke shell and execute your command inside the shell, its why it is called `execCommand`. This API call has corresponding std function https://github.com/nim-lang/Nim/blob/devel/lib/pure/osproc.nim#L99

Reply:
If you like to have arguments list and still execute your command via shell, you can use `startProcess(poEvalCommand)`.

Reply:
> > arguments.join
> 
> this would introduce quoting / whitespace issues, no?

This should be reviewed and i left appropriate TODOs in the code.

Reply:
> It doesnt look like inconsistency of the API, because `execCommand` is supposed to invoke shell and execute your command inside the shell, its why it is called `execCommand`. This API call has corresponding std function https://github.com/nim-lang/Nim/blob/devel/lib/pure/osproc.nim#L99

got it, thanks.